### ISSUE TITLE
模块里layernorm的实现疑问

### ISSUE THREAD

**Issue by HuiiJi** [megvii-research/NAFNet#170]
arch里 LayerNorm，更接近 InstanceNorm 或 GroupNorm 的变体，
mu = x.mean(1, keepdim=True)       # shape: (N, 1, H, W)
var = (x - mu).pow(2).mean(1, keepdim=True)  # shape: (N, 1, H, W)
计算维度：沿 C 计算均值和方差，即对每个样本的每个空间位置 (H, W) 独立计算 mu 和 var。而标准的layernorm是沿着CHW计算均值和方差，即每个样本的所有空间位置共享一套均值和方差；

self.weight = nn.Parameter(torch.ones(c))  # shape: (C,)
self.bias = nn.Parameter(torch.zeros(c))  # shape: (C,)
参数形状：(C,)，即 每个通道共享一组 gamma 和 beta。
y = self.weight.view(1, self.c, 1, 1) * y + self.bias.view(1, self.c, 1, 1)
gamma 和 beta 被广播到 (N, C, H, W)，但每个通道的所有 (H, W) 位置共享相同的缩放和偏移；而标准的layernorm的gamma和beta应该为（1，C，H，W），来对每个通道的每个空间位置调整。

因此个人理解代码里layernorm模块实际上更适应全局的风格迁移、像素修复等，不太适合局部像素处理？

**Comment by Colwzq** [comment]
这种mu和var的shape是(N,1,H,W)是第一次见，对每一个像素的所有通道做归一化看起来不太合理，如果两个通道提取到的信息是平坦和纹理，那么归一化之后的信息是很复杂的，我觉得反而不利于网络训练。

**Comment by Okitasandesu** [comment]
我认为arch里的LayerNorm反而跟InstanceNorm相去甚远，理解它们的方法差异很大，只不过它们的Affine参数都是C个，在通道维度上各异，在空间维度上共享

这个实现更像LayerNorm在序列问题（比如NLP）的应用，将每个像素作为一个token，对特征向量进行Normalization，然后对各特征分量进行各自的Affine。这也是普通的Transformer里LN的实现

我问过Gemini 2.5pro关于这种LN实现要如何理解，因为它确实和BatchNorm、InstanceNorm的表现不一样——BN、IN在对各个特征通道平面的数据分布进行Normalization和整体Affine，而这里的LN却是把各个像素的特征向量进行Normalization和逐元素Affine。Gemini 2.5pro给我的视角是：不要从统计学角度（分布）去理解，而是从向量空间的角度去理解。LN的Normalization把每个特征向量的L2长度限制为sqrt(C)，把特征分量的均值化为0；LN的Affine对各特征分量进行加权和偏移，加权是在调整通道的”重要性“（所以LN的Affine有点像通道注意力，只不过通道权重是从整个数据集里学到的，对每个样本都是相同的，但如果对每个样本都相同，这可能不符合注意力的本意，所以仍然不能算作注意力机制），偏移我不太理解，Gemini也只说是提高模型的表达能力

至于利不利于训练之类的实际效果问题，可能还是要实验才知道了，有没有做过实验的说一下呢？

**Comment by Kyekyekye** [comment]
我在gopro数据集尝试了将LayerNorm 改成BatchNorm（batchsize=40），其他参数均相同（lr，网络结构等），可以发现BN不利于网络训练。

<img width="717" height="301" alt="Image" src="https://github.com/user-attachments/assets/3cd6f10f-4e09-4ebf-828d-c80b6c100320" />

**Comment by Okitasandesu** [comment]
BatchNorm在许多图像到图像任务（超分、修复等）中确实已经很少用了，一个原因是它的性能和使用依赖batchsize（一般bs越大性能越好，bs=1时无法使用），另一个原因可能是它要求测试集和训练集的统计量要一致。例如，在低光图像增强中，训练集输入一般全是低光照的。如果要求网络在输入正常光照图像时也能正常工作（维持正常亮度，不过度增亮），BN是做不好的，因为测试时（正常光照）的统计量和训练时（低光照）的统计量（比如灰度均值）不同。此外，如果任务与某个统计量强相关（例如低光照与灰度均值），那么训练集的这个统计量会对BN的效果影响较大。例如在低光图像增强中，如果训练集里输入既有低光照也有正常光照，那么灰度均值就处于低光照和正常光照之间，这让网络在学习处理低光照图像时，可能比“训练集全是低光照，从而灰度均值是低光照”这种情况要困难。InstanceNorm、LayerNorm之类的都是基于当前样本的统计量，不记录整个数据集，因此不受数据集统计量的影响。
这可能是Deblur里BN比LN差的原因，此时无论bs多大都无法解决问题（不过这也只是我的猜测，如果把BN换成IN后效果有提升，那应该就是有这方面的原因）。

**Comment by Kyekyekye** [comment]
根据 @Okitasandesu 的猜想， 我快速验证了一下 LN，BN，IN之间的差异， 虽然还没训练完， 但是能看出在早期就已经有明显的分化（如果后期loss曲线有明显的变化，我会更新出来）

<img width="721" height="300" alt="Image" src="https://github.com/user-attachments/assets/9f424b56-dbcd-4615-82f1-47a60371e9d6" />

最终结果
<img width="627" height="301" alt="Image" src="https://github.com/user-attachments/assets/b5986aa6-1e2e-4fc9-8a3e-8b7b48c37604" />

但是IN的结果会出现artifact，和这里的相似 https://github.com/megvii-research/NAFNet/issues/25
